### PR TITLE
Driving improvements 🤩 

### PR DESCRIPTION
**Pull request by alicekuznetsov**
- Cardinal Snapping
- Snapping to the nearest chain
- Ability to zero your pose and gyro that ACTUALLY works on both alliances (ignore it not working in sim I'll fix that in supercore) 
- Slow Mode

### CODE REVIEW

File: src/main/java/frc/robot/commands/Drive.java [hunk 1]
```diff
@@ -34,23 +47,50 @@ public Drive(Drivetrain subDrivetrain, DoubleSupplier xAxis, DoubleSupplier yAxi
...
+          .in(Units.RadiansPerSecond);
+    } else if (south.getAsBoolean()) {
+      rVelocity = subDrivetrain.getVelocityToSnap(Units.Degrees.of(180 + northYaw.in(Units.Degrees)))
+          .in(Units.RadiansPerSecond);
+    } else if (west.getAsBoolean()) {
+      rVelocity = subDrivetrain.getVelocityToSnap(Units.Degrees.of(90 + northYaw.in(Units.Degrees)))
+          .in(Units.RadiansPerSecond);
+    } else if (chain.getAsBoolean()) {
+      rVelocity = subDrivetrain.getVelocityToChain().in(Units.RadiansPerSecond);
     }
```
Comment (TaylerUva) [inline src/main/java/frc/robot/commands/Drive.java]:
```suggestion
    } else if (north.getAsBoolean()) {
      rVelocity = subDrivetrain.getVelocityToSnap(northYaw.plus(Units.Degrees.of(0)));
```
rVelocity should be a velocity object to make all this math and code simpler. we should always convert at the end. and eventually we will be updating SuperCORE to support velocity natively

**Review by TaylerUva — CHANGES_REQUESTED** [review]
see comments

**Comment by alicekuznetsov** [comment]
Should be all good to merge in now!
Feel free to test on the bot if needed for sanity checks but I did finally fix our sim (https://github.com/FRCTeam3255/SuperCORE/pull/118) so we should be fine

File: src/main/java/frc/robot/RobotContainer.java [hunk 2]
```diff
@@ -37,9 +40,9 @@ public RobotContainer() {
 
   private void configureBindings() {
     conDriver.btn_B.onTrue(Commands.runOnce(() -> subDrivetrain.resetModulesToAbsolute()));
-    conDriver.btn_Start.onTrue(Commands.runOnce(() -> subDrivetrain.resetYaw(180)));
+
     conDriver.btn_Back.onTrue(
-        Commands.runOnce(() -> subDrivetrain.resetPoseToPose(new Pose2d(1.35, 5.50, Rotation2d.fromDegrees(180)))));
+        Commands.runOnce(() -> subDrivetrain.resetPoseToPose(constField.getFieldPositions().get()[6].toPose2d())));
```
Comment (TaylerUva) [inline src/main/java/frc/robot/RobotContainer.java:45]: I think we need to bump the SuperCORE version to use this, this way

File: src/main/java/frc/robot/commands/Drive.java [hunk 3]
```diff
@@ -34,23 +50,45 @@ public Drive(Drivetrain subDrivetrain, DoubleSupplier xAxis, DoubleSupplier yAxi
 
   @Override
   public void initialize() {
+    northYaw = constField.isRedAlliance() ? Units.Degrees.of(180) : Units.Degrees.of(0);
+    redAllianceMultiplier = constField.isRedAlliance() ? -1 : 1;
```
Comment (TaylerUva) [inline src/main/java/frc/robot/commands/Drive.java:54]: What is this for?

**Review by TaylerUva — COMMENTED** [review]
Looks good but had some questions. Should also pull from main

File: src/main/java/frc/robot/commands/Drive.java (hunk 3 above)
Comment (alicekuznetsov) [reply src/main/java/frc/robot/commands/Drive.java:54]: The direction that is up/north changes depending on the alliance, so we need to multiply some values by -1 which is established here
Unrelated, but we should probably put all of this in the swerve template as well + an example auto

File: src/main/java/frc/robot/RobotContainer.java (hunk 2 above)
Comment (alicekuznetsov) [reply src/main/java/frc/robot/RobotContainer.java:45]: yes yes yes this does need to happen

**Review by TaylerUva — APPROVED** [review]

**Pull request merged**